pkg/operator/ceph/object/shared_pools.go

```diff
@@ -0,0 +1,532 @@
+	}
+	zoneArg := fmt.Sprintf("--rgw-zone=%s", objContext.Zone)
+
+	if objContext.ZoneGroup == "" {
+		logger.Warning("update zonegroup: object store zonegroup is missing from context")
+	}
+	zoneGroupArg := fmt.Sprintf("--rgw-zonegroup=%s", objContext.ZoneGroup)
+
+	configBytes, err := group.toJson()
+	if err != nil {
+		return nil, err
+	}
+	configFilename := path.Join(objContext.Context.ConfigDir, objContext.Name+".zonegroupcfg")
+	if err := os.WriteFile(configFilename, configBytes, 0600); err != nil {
+		return nil, errors.Wrap(err, "failed to write zonegroup config file")
+	}
+	defer os.Remove(configFilename)
+
+	args := []string{"zonegroup", "set", zoneArg, "--infile=" + configFilename, realmArg, zoneGroupArg}
+	updatedBytes, err := RunAdminCommandNoMultisite(objContext, false, args...)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set zone config")
+	}
+	updated := ZoneGroupJSON{}
+	err = json.Unmarshal([]byte(updatedBytes), &updated)
+	return updated, err
+}
+
+type ZonegroupPlacementTarget struct {
+	Name           string        `json:"name"`
+	StorageClasses []string      `json:"storage_classes"`
+	Tags           []interface{} `json:"tags"`
+}
+
+type ZonePlacementPool struct {
+	Key string               `json:"key"`
+	Val ZonePlacementPoolVal `json:"val"`
+}
+
+type ZonePlacementPoolVal struct {
+	DataExtraPool  string                               `json:"data_extra_pool"`
+	IndexPool      string                               `json:"index_pool"`
+	IndexType      int                                  `json:"index_type"`
+	InlineData     bool                                 `json:"inline_data"`
+	StorageClasses map[string]ZonePlacementStorageClass `json:"storage_classes"`
+}
+
+type ZonePlacementStorageClass struct {
+	DataPool string `json:"data_pool"`
+}
+
+type ZoneJSON map[string]json.RawMessage
+
+func (z ZoneJSON) getPoolPlacements() ([]ZonePlacementPool, error) {
+	placementsJSON, ok := z["placement_pools"]
+	if !ok {
+		return nil, nil
+	}
+	res := []ZonePlacementPool{}
+	err := json.Unmarshal(placementsJSON, &res)
+	return res, err
+}
+
+func (z ZoneJSON) setPoolPlacements(pp []ZonePlacementPool) error {
+	bytes, err := json.Marshal(&pp)
+	if err != nil {
+		return err
+	}
+	z["placement_pools"] = json.RawMessage(bytes)
+	return nil
+}
+
+func (z ZoneJSON) getString(field string) (string, error) {
+	raw, ok := z[field]
+	if !ok {
+		return "", nil
+	}
+	var res string
+	return res, json.Unmarshal(raw, &res)
+}
+
+func (z ZoneJSON) setString(field, val string) {
+	z[field] = json.RawMessage(fmt.Sprintf("%q", val))
+}
+
+func (z ZoneJSON) toJson() ([]byte, error) {
+	return json.Marshal(&z)
+}
+
+type ZoneGroupJSON map[string]json.RawMessage
+
+func (z ZoneGroupJSON) getPlacementTargets() ([]ZonegroupPlacementTarget, error) {
+	targetsJSON, ok := z["placement_targets"]
+	if !ok {
+		return nil, nil
+	}
+	res := []ZonegroupPlacementTarget{}
+	err := json.Unmarshal(targetsJSON, &res)
+	return res, err
+}
```

Comment:
[Old code](https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/objectstore.go#L918-L944) was more straightforfard because it was mostly working with json maps, not arrays. You can see that it just always takes the first element to update placement.

I can try to continue using `map[string]interface{}` and iterate over placements bet there will be a lot of boilerplate code to access deeply nested properties.